contracts/Enforcer.sol

```diff
@@ -29,64 +29,79 @@ contract Enforcer {
         _;
     }
 
-    event Registered(bytes32 indexed _executionId, address indexed _solver, bytes _code, bytes _callData);
-    event Slashed(bytes32 indexed _executionId, address indexed _address);
+    event Registered(bytes32 indexed executionId, address indexed solver, address codeContractAddress, bytes _callData);
+    event DisputeInitialised(bytes32 indexed disputeId, bytes32 indexed executionId);
+    event Slashed(bytes32 indexed executionId, address indexed _address);
 
     constructor(address _verifier, uint256 _challengePeriod, uint256 _bondAmount) public {
-        verifier = IVerifier(_verifier);
+        verifier = Verifier(_verifier);
         challengePeriod = _challengePeriod;
         bondAmount = _bondAmount;
     }
 
     // register a new execution
-    function register(bytes memory _code, bytes memory _callData, bytes32 _endHash, uint256 _executionLength)
-        public payable returns(bytes32 executionId)
+    function register(address codeContractAddress, bytes memory _callData, bytes32 endHash, uint256 executionLength)
+        public payable
     {
         require(msg.value == bondAmount);
-        executionId = keccak256(abi.encodePacked(_code, _callData));
+
+        bytes32 executionId = keccak256(abi.encodePacked(codeContractAddress, _callData));
+
         require(executions[executionId].startBlock == 0);
-        executions[executionId] = Execution(block.number, _endHash, _executionLength, msg.sender);
+        executions[executionId] = Execution(block.number, endHash, executionLength, msg.sender);
         bonds[msg.sender] += bondAmount;
-        emit Registered(executionId, msg.sender, _code, _callData);
+
+        emit Registered(executionId, msg.sender, codeContractAddress, _callData);
     }
 
     // starts a new dispute
-    function dispute(bytes32 _executionId, bytes32 _endHash, uint256 _executionLength) public payable {
-        Execution storage execution = executions[_executionId];
+    function dispute(address codeContractAddress, bytes memory _callData, bytes32 endHash)
+        public payable
+    {
+        bytes32 executionId = keccak256(abi.encodePacked(codeContractAddress, _callData));
+
+        Execution storage execution = executions[executionId];
+
         require(execution.startBlock != 0);
         require(execution.startBlock + challengePeriod > block.number);
         require(msg.value == bondAmount);
-        require(execution.endHash != _endHash);
+        // Do we want to prohibit early?
+        // require(execution.endHash != endHash);
```

Comment:
I commented that out. I still use this case in the verifier tests. The solver wins in this case at the end. 
@johannbarbie Thoughts?